pkg/storage/unified/resource/datastore.go

```diff
@@ -0,0 +1,151 @@
+package resource
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"iter"
+	"strconv"
+	"strings"
+)
+
+const (
+	dataSection = "unified/data"
+)
+
+// dataStore is a data store that uses a KV store to store data.
+type dataStore struct {
+	kv KV
+}
+
+func newDataStore(kv KV) *dataStore {
+	return &dataStore{
+		kv: kv,
+	}
+}
+
+type DataObj struct {
+	Key   DataKey
+	Value io.ReadCloser
+}
+
+type DataKey struct {
+	Namespace       string
+	Group           string
+	Resource        string
+	Name            string
+	ResourceVersion int64
+	Action          DataAction
+}
+
+type ListRequestKey struct {
+	Namespace string
+	Group     string
+	Resource  string
+	Name      string
+}
+
+type DataAction string
+
+const (
+	DataActionCreated DataAction = "created"
+	DataActionUpdated DataAction = "updated"
+	DataActionDeleted DataAction = "deleted"
+)
+
+// Keys returns all keys for a given key by iterating through the KV store
+func (d *dataStore) Keys(ctx context.Context, key ListRequestKey) iter.Seq2[DataKey, error] {
+	prefix, err := d.getPrefix(key)
+	if err != nil {
+		return func(yield func(DataKey, error) bool) {
+			yield(DataKey{}, err)
+		}
+	}
+	return func(yield func(DataKey, error) bool) {
+		for k, err := range d.kv.Keys(ctx, dataSection, ListOptions{
+			StartKey: prefix,
+			EndKey:   PrefixRangeEnd(prefix),
+		}) {
+			if err != nil {
+				yield(DataKey{}, err)
+				return
+			}
+			key, err := d.parseKey(k)
+			if err != nil {
+				yield(DataKey{}, err)
+				return
+			}
+			if !yield(key, nil) {
+				return
+			}
+		}
+	}
+}
+
+func (d *dataStore) Get(ctx context.Context, key DataKey) (io.ReadCloser, error) {
+	obj, err := d.kv.Get(ctx, dataSection, d.getKey(key))
+	if err != nil {
+		return nil, err
+	}
+	return obj.Value, nil
+}
+
+func (d *dataStore) Save(ctx context.Context, key DataKey, value io.ReadCloser) error {
+	return d.kv.Save(ctx, dataSection, d.getKey(key), value)
```

Comment:
Shall we verify that all parts of the key are set here to avoid creating bad keys? (Perhaps `validate` function on `DataKey`)